Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1289 @@
+		}
+		else
+		{
+			// If the last one, move directly
+			m_pMyUpgradeSLot[iDestiOrder] = pSrc;
+			m_pMyUpgradeInv[iSourceOrder] = nullptr;
+			char szID[32];
+			sprintf(szID, "a_m_%d", iDestiOrder);
+			CN3UIArea* pSlotArea = (CN3UIArea*)GetChildByID(szID);
+			SetupIconArea(pSrc, pSlotArea);
+		}
+	}
+	else
+	{
+		// If is not countable item, just move it
+		m_pMyUpgradeSLot[iDestiOrder] = pSrc;
+		m_pMyUpgradeInv[iSourceOrder] = nullptr;
+		char szID[32];
+		sprintf(szID, "a_m_%d", iDestiOrder);
+		CN3UIArea* pSlotArea = (CN3UIArea*)GetChildByID(szID);
+		SetupIconArea(pSrc, pSlotArea);
+	}
+	m_iUpgradeSlotInvPos[iDestiOrder+1] = iSourceOrder;
+	return true;
+}
+
+void CUIItemUpgrade::FlipFlopAnim()
+{
+	if (m_eAnimationState != ANIM_FLIPFLOP)
+		return;
+
+	// Hide before frame
+	if (m_iCurrentFrame > 0)
+	{
+		char szID[32];
+		sprintf(szID, m_bUpgradeSuccesfull ? "img_s_load_%d" : "img_f_load_%d", m_iCurrentFrame - 1);
+		if (CN3UIImage* pImg = (CN3UIImage*)GetChildByID(szID))
+			pImg->SetVisible(false);
+	}
+
+	// Show current frame
+	char szID[32];
+	sprintf(szID, m_bUpgradeSuccesfull ? "img_s_load_%d" : "img_f_load_%d", m_iCurrentFrame);
+	if (CN3UIImage* pImg = (CN3UIImage*) GetChildByID(szID))
+	{
+		pImg->SetVisible(true);
+		pImg->SetParent(this);
+	}
+
+}
+void CUIItemUpgrade::AnimClose()
+{
+	if (m_eAnimationState != ANIM_NONE)
+	{
+		float m_fAnimationTimer = 0.0f;
+		int m_iCurrentFrame = 0;
+		m_pImageCover1->SetVisible(false);
+		m_pImageCover2->SetVisible(false);
+	}
+
+	m_eAnimationState = ANIM_NONE;
+	HideAllAnimationFrames();
+}
+
+void CUIItemUpgrade::ShowResultUpgrade()
+{
+	// Show result item in result area if upgrade was successful
+	if (m_bUpgradeSuccesfull && m_pUpgradeResultSlot && m_pAreaResult)
+	{
+		// Set icon position to result area
+		m_pUpgradeResultSlot->pUIIcon->SetRegion(m_pAreaResult->GetRegion());
+		m_pUpgradeResultSlot->pUIIcon->SetMoveRect(m_pAreaResult->GetRegion());
+		m_pUpgradeResultSlot->pUIIcon->SetParent(this);
+		m_pUpgradeResultSlot->pUIIcon->SetVisible(true);
+	}
+
+	m_eAnimationState = ANIM_COVER_OPENING;
+}
+
+void CUIItemUpgrade::StartUpgradeAnim()
+{
+	if (m_bReceivedResultFromServer)
+	{
+		if (!m_pImageCover1 || !m_pImageCover2)
+			return;
+		m_fAnimationTimer = 0.0f;
+		m_iCurrentFrame = 0;
+
+		m_pImageCover1 = (CN3UIImage*) (this->GetChildByID("img_cover_01"));
+		__ASSERT(m_pImageCover1, "NULL UI Component!!");
+		m_pImageCover2 = (CN3UIImage*) (this->GetChildByID("img_cover_02"));
+		__ASSERT(m_pImageCover2, "NULL UI Component!!");
+
+		// Make covers visible during animation
+		m_pImageCover1->SetVisible(true);
+		m_pImageCover2->SetVisible(true);
+		m_pImageCover1->SetParent(this);
+		m_pImageCover2->SetParent(this);
+
+		// save original positions
+		m_rcCover1Original = m_pImageCover1->GetRegion();
+		m_rcCover2Original = m_pImageCover2->GetRegion();
+
+		m_eAnimationState = ANIM_FLIPFLOP;
+	}
+}
+
+void CUIItemUpgrade::HandleInventoryIconRightClick()
```

Comment:
This logic should directly influence `MouseProc()` as it needs to know if there was a successful interaction.
As such, we should at least be returning something here - like a boolean.